### PR TITLE
Experiment framework

### PR DESCRIPTION
**Pull request by Peterburnett**
Added forms and database tables for setup of experiments. No multiple condition support yet

### CODE REVIEW

File: classes/experimentmanager.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * @package   tool_abconfig
+ * @author    Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace tool_abconfig;
+
+defined('MOODLE_INTERNAL') || die();
+require_once(__DIR__.'/../../../config.php');
```
Comment (brendanheywood) [inline classes/experimentmanager.php]: config should never be included in classes, only ever in entry points ie cli, web

File: classes/form/edit_experiment.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    public function definition() {
+        $mform = $this->_form;
+
+        // Hidden form element for experiment id
+        $mform->addElement('hidden', 'shortname', $this->_customdata['shortname']);
+        $mform->setType('shortname', PARAM_TEXT);
+
+        // Experiment data
+        $name = $this->_customdata['name'];
```
Comment (brendanheywood) [inline classes/form/edit_experiment.php]: These 6 lines with customdata seem redundant, there should be a $mform->set_data($toform); call on the page which creates the form object which does all this in a single hit

File: classes/form/edit_experiment.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        $mform->setType('ipwhitelist', PARAM_TEXT);
+        $mform->setDefault('experimentipwhitelist', $ipwhitelist);
+        // Commands field
+        $mform->addElement('textarea', 'experimentcommands', get_string('formexperimentcommands', 'tool_abconfig'));
+        $mform->setType('experimentcommands', PARAM_RAW);
+        $mform->setDefault('experimentcommands', $commands);
+
+        // Value field
+        $mform->addElement('text', 'experimentvalue', get_string('formexperimentvalue', 'tool_abconfig'));
+        $mform->setType('experimentvalue', PARAM_TEXT);
```
Comment (brendanheywood) [inline classes/form/edit_experiment.php:69]: should this be an int?

File: classes/form/manage_experiments.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+        $shortname = $data['experimentshortname'];
+        $sqlexperiment = $DB->sql_compare_text($shortname, strlen($shortname));
+        $record = $DB->get_record_sql('SELECT * FROM {tool_abconfig_experiments} WHERE shortname = ?', array($sqlexperiment));
+
+        if (!empty($record)) {
+            $errors['experimentshortname'] = get_string('formexperimentalreadyexists', 'tool_abconfig');
+        }
+
+        $experiments = $DB->get_records('tool_abconfig_experiments');
```
Comment (brendanheywood) [inline classes/form/manage_experiments.php]: typo?

File: db/install.xml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        <FIELD NAME="scope" TYPE="text" NOTNULL="false" SEQUENCE="false"/>
+      </FIELDS>
+      <KEYS>
+        <KEY NAME="primary" TYPE="primary" FIELDS="id"/>
+      </KEYS>
+    </TABLE>
+    <TABLE NAME="tool_abconfig_conditions" COMMENT="Table to store conditions for experiments">
+      <FIELDS>
+        <FIELD NAME="id" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="true"/>
+        <FIELD NAME="experiment" TYPE="text" NOTNULL="false" SEQUENCE="false" COMMENT="Experiment shortname that the conditions relate to"/>
```
Comment (brendanheywood) [inline db/install.xml]: this should be an int and have a foreign key back to experiment

File: db/install.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<XMLDB PATH="admin/tool/abconfig/db" VERSION="20190829" COMMENT="XMLDB file for Moodle admin/tool/abconfig"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:noNamespaceSchemaLocation="../../../../lib/xmldb/xmldb.xsd"
+>
+  <TABLES>
+    <TABLE NAME="tool_abconfig_experiments" COMMENT="Table to store experiments">
```
Comment (brendanheywood) [inline db/install.xml]: hmm can't find the right wiki doc but 90% sure there is a convention to always use the singular for tables names ie 'experiment' without the s

File: db/install.xml (hunk 5 above)
Comment (brendanheywood) [reply db/install.xml]: as above tool_abconfig_condition

vs

tool_abconfig_conditions

File: edit_experiment.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    $shortname = $fromform->shortname;
+    $iplist = $fromform->experimentipwhitelist;
+    $commands = $fromform->experimentcommands;
+    $value = $fromform->experimentvalue;
+
+    $sqlconditions = $DB->sql_compare_text($shortname, strlen($shortname));
+    $record = $DB->get_record_sql('SELECT * FROM {tool_abconfig_conditions} WHERE experiment = ?', array($sqlconditions));
+
+    // If record doesnt exist, create record, else, update record
+    if (empty($record)) {
```
Comment (brendanheywood) [inline edit_experiment.php:92]: all this logic looks to assume that there is only 1 condition per experiment? we need to support multiple

File: settings.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    $ADMIN->add('abconfig', new admin_externalpage('tool_abconfig_manageexperiments',
+    get_string('manageexperimentspagename', 'tool_abconfig'),
+    new moodle_url('/admin/tool/abconfig/manage_experiments.php')));
+
+    $settings = new admin_settingpage('abconfigsettings', get_string('abconfigsettings', 'tool_abconfig'));
+    $ADMIN->add('abconfig', $settings);
+
+    if (!during_initial_install()) {
+
+        $settings->add(new admin_setting_configcheckbox('tool_abconfig/enable_plugin', get_string('settingsenablename', 'tool_abconfig'),
```
Comment (brendanheywood) [inline settings.php]: Do we need this setting at all? We already have the core plugin enable flag, and each experiment will have it's own, I think this not needed

File: edit_experiment.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+require_login();
+
+global $DB, $PAGE, $SESSION;
+$prevurl = ($CFG->wwwroot.'/admin/tool/abconfig/manage_experiments.php');
+
+$eid = optional_param('id', 0, PARAM_INT);
+
+// store eid if set in params (for page submission and refresh)
+if ($eid != 0) {
+    $SESSION->eid = $eid;
```
Comment (brendanheywood) [inline edit_experiment.php]: why do we need to store anything is the session?

File: classes/form/edit_experiment.php (hunk 3 above)
Comment (Peterburnett) [reply classes/form/edit_experiment.php:69]: If it is set to PARAM_INT, it converts most text to 0, which may be an unintended effect. This way, it is checked for is_numeric in the validation, and then an error is returned to the user

**Pull request merged**